pubsub/alicloud/mns/mns.go

```diff
@@ -0,0 +1,303 @@
+			m.logger.Error(err)
+			return err
+		}
+
+		topic := ali_mns.NewMNSTopic(req.Topic, m.client)
+		topic.PublishMessage(msgPublishReq)
+
+	case MnsModeQueue:
+		queue := ali_mns.NewMNSQueue(metaData.QueueName, m.client)
+		_, err := queue.SendMessage(msgSendReq)
+		if err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("unsupported MNS mode: %v, should be queue or topic", m.settings.MnsMode)
+	}
+
+	return nil
+}
+
+// Subscribe
+func (m *mns) Subscribe(req pubsub.SubscribeRequest, handler pubsub.Handler) error {
+	var metaData RequestMetaData
+	metaData.Decode(req.Metadata)
+
+	// create/fetch queue
+	err := m.queueManager.CreateQueue(
+		metaData.QueueName,
+		metaData.QueueDelaySeconds,
+		metaData.QueueMaxMessageSize,
+		metaData.QueueMessageRetentionPeriod,
+		metaData.QueueVisibilityTimeout,
+		metaData.QueuePollingWaitSeconds,
+		metaData.QueueSlices,
+	)
+	if err != nil && !ali_mns.ERR_MNS_QUEUE_ALREADY_EXIST_AND_HAVE_SAME_ATTR.IsEqual(err) {
+		m.logger.Error(err)
+		return err
+	}
+
+	switch m.settings.MnsMode {
+	case MnsModeTopic:
+		// create/fetch topic
+		err = m.topicManager.CreateTopic(
+			req.Topic,
+			metaData.TopicMaxMessageSize,
+			metaData.TopicLoggingEnabled,
+		)
+		if err != nil && !ali_mns.ERR_MNS_TOPIC_ALREADY_EXIST_AND_HAVE_SAME_ATTR.IsEqual(err) {
+			m.logger.Error(err)
+			return err
+		}
+
+		// subscribe topic
+		topic := ali_mns.NewMNSTopic(req.Topic, m.client)
+		sub := ali_mns.MessageSubsribeRequest{
+			Endpoint:            topic.GenerateQueueEndpoint(metaData.QueueName),
+			NotifyContentFormat: metaData.SubscriptionNotifyContentFormat,
+		}
+
+		err = topic.Subscribe(metaData.SubscriptionName, sub)
+		if err != nil && !ali_mns.ERR_MNS_SUBSCRIPTION_ALREADY_EXIST_AND_HAVE_SAME_ATTR.IsEqual(err) {
+			m.logger.Error(err)
+			return err
+		}
+
+		time.Sleep(time.Duration(2) * time.Second)
+	case MnsModeQueue:
+		// do nothing here because queue is already created or fetched
+	default:
+		return fmt.Errorf("unsupported MNS mode: %v, should be queue or topic", m.settings.MnsMode)
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	queue := ali_mns.NewMNSQueue(metaData.QueueName, m.client)
+
+	respChan := make(chan ali_mns.MessageReceiveResponse)
+	errChan := make(chan error)
+
+	go m.processMessageLoop(ctx, req.Topic, queue, handler, respChan, errChan)
+	go m.receiveMessageLoop(ctx, queue, respChan, errChan)
+
+	switch m.settings.MnsMode {
+	case MnsModeQueue:
+		if c, ok := m.queueContexts[metaData.QueueName]; ok {
+			c.cancel()
+		}
+		m.queueContexts[metaData.QueueName] = contextWithCancelFunc{ctx, cancelFunc}
+	case MnsModeTopic:
+		if c, ok := m.subscriptionContexts[metaData.SubscriptionName]; ok {
+			c.cancel()
+		}
+		m.subscriptionContexts[metaData.SubscriptionName] = contextWithCancelFunc{ctx, cancelFunc}
+	}
+
+	return nil
+}
+
+func wrapMnsMessage(resp ali_mns.MessageReceiveResponse, topic string) (msg pubsub.NewMessage, err error) {
```

Comment:
Nit: Go naming `wrapMnsMessage` -> `wrapMNSMessage` (assuming MNS is an acronym)